docsource/ops.rst

```diff
@@ -1,6 +1,236 @@
+
+Reindexing Process
+-------------------
+For the sake of simplicity and for the reasons described below, a single process has been defined for all reindexing cases -
+
+a) change field value
+b) add field type
+c) change field type
+d) remove field type
+
+The process involves reindexing an existing index into a new index for all of these cases. i.e. in-place reindexing is discouraged because the system may not have a recent snapshot/backup or any snapshot for recovery purposes should anything go wrong with the reindexing process.
+
+In addition, given that there can be as many as a thousand indices for KBase data, maintenance can become a challenge if the process is not simple. Some level of simplicity has been achieved here by defiining a single process that covers all the reindexing cases. If necessary, the process may be further simplified through some level of automation as it matures over time.
+
+.. note::
+
+    The commands below can be copy-pasted into Kibana and executed against the index. The corresponding curl commands can be obtained from Kibana by clicking on the little wrench icon that appears next to the pasted command.
+
+.. note::
+
+    If any of the steps below fail, don't proceed until the issue is resolved by referring to the ElasticSearch documentation.
+
+1. Note current time
+
+2. Stop any workers that are performing trickle updates on the index that needs to be reindexed.
+
+3. Refresh the index that needs reindexing to make sure it has been brought to a consistent state.
+
+.. code-block:: bash
+
+    POST /kbase.genome_1/_refresh
+
+4. Get a checksum for the index and record it in a separate file for later verification.
+
+.. code-block:: bash
+
+    GET /kbase.genome_1/_stats/docs,store
+
+5a. If mapping needs to be changed (for cases b, c, d above), get current index mapping.
+
+.. code-block:: bash
+
+    GET kbase.genome_1/_mapping
+
+5b. Copy-paste the mapping from the current index into the body section of the PUT command below and make the necessary field change (preferably one change per complete reindexing operation).
+
+It is a good practice to make the mapping strict ("dynamic": "strict") for each type (data and access) in the index. Strict mappings prevent the mapping from being modified dynamically during ingest time.
+
+Update the settings section below the mapping. The number of shards and replicas must be decided on based on your capacity planning rules. It is costly to change the number of shards, so choose wisely! Make sure not to exceed 600 shards for any node in the system. Increase number of replicas to improve availability. In general,
```

Comment:
It's not relevant to reindexing but it is important to have a separate capacity plan as it involves the determination of the number of ingest nodes, data nodes etc among other things. I think we should take this up as a separate task and work it out based on current needs and our projected growth. Since a lot of decision making in ES starts with "it depends" it might be better to start with defaults, see how the system responds and then use that baseline at least to determine current needs. I have documented limits here for #shards and shard size as a means to double check if we are within reasonable bounds.

Reply:
> It's not relevant to reindexing

Naively it seems like it is since you'll have to create a new index and you'll want to create it with the correct number of shards.